File: 2025-dev-week2-building-a-p2sh-p2wsh-multisig-tx/python/main.py
import hashlib
import binascii
import ecdsa
import base58
# from bitcoinlib.keys import Key
# from bitcoinlib.transactions import Transaction
# from bitcoinlib.scripts import Script

def solve():
    # these are the constants given in the readme file
    priv_key1 = "39dc0a9f0b185a2ee56349691f34716e6e0cda06a7f9707742ac113c4e2317bf"
    priv_key2 = "5077ccd9c558b7d04a81920d38aa11b4a9f9de3b23fab45c3ef28039920fdd6d"
    redeem_script_hex = "5221032ff8c5df0bc00fe1ac2319c3b8070d6d1e04cfbf4fedda499ae7b775185ad53b21039bbc8d24f89e5bc44c5b0d1980d6658316a6b2440023117c3c03a4975b04dd5652ae"

    # converting the redeem_script_hex to bytes
    redeem_script = bytes.fromhex(redeem_script_hex)
    # caclculating the witness_program
    witness_program = hashlib.sha256(redeem_script).digest()
    # Create p2sh redeem script (0x00 + 0x20 + witness_program)
    p2sh_redeem_script = b"\x00\x20" + hashlib.sha256(redeem_script).digest()

    ripemd160 = hashlib.new('ripemd160')
    ripemd160.update(hashlib.sha256(p2sh_redeem_script).digest())
    p2sh_hash = ripemd160.digest()

    # Transaction details that were mentioned in the readme file
    version = (1).to_bytes(4, byteorder='little')
    input_count = (1).to_bytes(1, byteorder='little')
    prev_txid = bytes.fromhex("0000000000000000000000000000000000000000000000000000000000000000")
    prev_index = (0).to_bytes(4, byteorder='little')
    sequence = (0xffffffff).to_bytes(4, byteorder='little')
    output_count = (1).to_bytes(1, byteorder='little')
    value = (100000).to_bytes(8, byteorder='little')
    address = "325UUecEQuyrTd28Xs2hvAxdAjHM7XzqVF"
    addr_bytes = base58.b58decode_check(address)
    script_hash = addr_bytes[1:]
    output_script = b"\xa9\x14" + script_hash + b"\x87"
    script_len = len(output_script).to_bytes(1, byteorder='little')
    locktime = (0).to_bytes(4, byteorder='little')

    # Calculating the script signature
    script_sig = len(p2sh_redeem_script).to_bytes(1, byteorder='little') + p2sh_redeem_script
    script_sig_len = len(script_sig).to_bytes(1, byteorder='little')
    hash_prevouts = hashlib.sha256(hashlib.sha256(prev_txid + prev_index).digest()).digest()
    hash_sequence = hashlib.sha256(hashlib.sha256(sequence).digest()).digest()
    hash_outputs = hashlib.sha256(hashlib.sha256(value + script_len + output_script).digest()).digest()

    # amount as mentioned in the readme file
    amount = (100000).to_bytes(8, byteorder='little')

    sighash_preimage = (
        version + hash_prevouts + hash_sequence +
        prev_txid + prev_index +
        len(redeem_script).to_bytes(1, byteorder='little') + redeem_script +
        amount + sequence + hash_outputs + locktime + (1).to_bytes(4, byteorder='little')
    )

    sighash = hashlib.sha256(hashlib.sha256(sighash_preimage).digest()).digest()

    # signature of the primary keys
    sk1 = ecdsa.SigningKey.from_string(bytes.fromhex(priv_key1), curve=ecdsa.SECP256k1)
    sk2 = ecdsa.SigningKey.from_string(bytes.fromhex(priv_key2), curve=ecdsa.SECP256k1)

    sig1 = sk1.sign_digest(sighash, sigencode=ecdsa.util.sigencode_der_canonize) + b"\x01"
    sig2 = sk2.sign_digest(sighash, sigencode=ecdsa.util.sigencode_der_canonize) + b"\x01"

    # the public keys
    pubkey1 = redeem_script[2:35]
    pubkey2 = redeem_script[36:69]

    # an empty byte so as to cancel out the op_multisig
    # sig2 and 1 in reverse order since this was the order in the redeem script
    witness = [b"", sig2, sig1, redeem_script]

    witness_encoded = len(witness).to_bytes(1, byteorder='little') + b"".join(
        len(item).to_bytes(1, byteorder='little') + item for item in witness)

    # final transaction id
    final_tx = (
        version + b"\x00\x01" + input_count +
        prev_txid + prev_index + script_sig_len + script_sig + sequence +
        output_count + value + script_len + output_script +
        witness_encoded + locktime
    )

    # converting it to hex
    tx_hex = binascii.hexlify(final_tx).decode()
    print(tx_hex)

    file = open("out.txt","w")
    file.write(tx_hex)
    file.close()

    # redeem_script_hex = "5221032ff8c5df0bc00fe1ac2319c3b8070d6d1e04cfbf4fedda499ae7b775185ad53b21039bbc8d24f89e5bc44c5b0d1980d6658316a6b2440023117c3c03a4975b04dd5652ae"
    # redeem_script = bytes.fromhex(redeem_script_hex)
    
    # witness_hash = hashlib.sha256(redeem_script).digest()
    # witness_program = bytes.fromhex("0020") + witness_hash

    # unlocking_script = Script([witness_program]).as_bytes()    
    # locking_script = Script([b'\xa9', bytes(hashlib.new('ripemd160', witness_hash).digest()), b'\x87']).as_bytes()
    
    # txid = "0000000000000000000000000000000000000000000000000000000000000000"
    # vout = 0
    # sequence = 0xffffffff
    # input_value = 100000
    
    # expected_address = "325UUecEQuyrTd28Xs2hvAxdAjHM7XzqVF"
    # output_value = 100000
    # locktime = 0

    # tx = Transaction(version=2,locktime=locktime,network='bitcoin') # here network idk what the use is , network='bitcoin'
    # tx.add_input(
    #     prev_txid=txid,
    #     output_n=vout,
    #     sequence=sequence,
    #     unlocking_script=unlocking_script,
    #     # address=expected_address,
    #     witness_type="p2sh-segwit",
    #     value=input_value,
    #     locking_script=locking_script
    # )
    # tx.add_output(output_value, expected_address)
    

    # privkey1_hex = "39dc0a9f0b185a2ee56349691f34716e6e0cda06a7f9707742ac113c4e2317bf"
    # privkey2_hex = "5077ccd9c558b7d04a81920d38aa11b4a9f9de3b23fab45c3ef28039920fdd6d"
    # key1 = Key(import_key=privkey1_hex)
    # key2 = Key(import_key=privkey2_hex)
    # sk1 = SigningKey.from_string(key1.private_byte, curve=SECP256k1)
    # sk2 = SigningKey.from_string(key2.private_byte, curve=SECP256k1)
    # sighash_bytes = tx.signature_hash(0, hash_type=1, witness_type='segwit', as_hex=False)
    # sig1 = sk1.sign_digest_deterministic(sighash_bytes, sigencode=util.sigencode_der) + b'\x01'
    # sig2 = sk2.sign_digest_deterministic(sighash_bytes, sigencode=util.sigencode_der) + b'\x01'
    
    # tx.inputs[0].witnesses = [b'', sig2, sig1, redeem_script]
    
    # raw_tx_hex = tx.raw_hex()
    
    # print("Raw tx hex:", hash(raw_tx_hex))
    # with open("out.txt", "w", encoding="utf-8") as f:
    #     f.write(raw_tx_hex)
if __name__=="__main__":
    solve()

File: 2025-dev-week2-building-a-p2sh-p2wsh-multisig-tx/rust/src/main.rs
use sha2::{Sha256, Digest};
use ripemd::Ripemd160;
use secp256k1::{Secp256k1, SecretKey, Message};
use std::fs::File;
use std::io::Write;
use bs58;
use hex;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Private keys and redeem script from the assignment
    let priv_key1 = "39dc0a9f0b185a2ee56349691f34716e6e0cda06a7f9707742ac113c4e2317bf";
    let priv_key2 = "5077ccd9c558b7d04a81920d38aa11b4a9f9de3b23fab45c3ef28039920fdd6d";
    let redeem_script_hex = "5221032ff8c5df0bc00fe1ac2319c3b8070d6d1e04cfbf4fedda499ae7b775185ad53b21039bbc8d24f89e5bc44c5b0d1980d6658316a6b2440023117c3c03a4975b04dd5652ae";

    // Convert hex to bytes
    let redeem_script = hex::decode(redeem_script_hex)?;

    // Calculate the witness program (SHA256 of redeem script)
    let mut hasher = Sha256::new();
    hasher.update(&redeem_script);
    let witness_program = hasher.finalize();

    // Create P2SH redeem script (0x00 + 0x20 + witness_program)
    let mut p2sh_redeem_script = Vec::new();
    p2sh_redeem_script.push(0x00); // Version 0 (P2WSH)
    p2sh_redeem_script.push(0x20); // 32 bytes (SHA256 hash length)
    p2sh_redeem_script.extend_from_slice(&witness_program);

    // Hash the P2SH redeem script to get P2SH address
    let mut hasher = Sha256::new();
    hasher.update(&p2sh_redeem_script);
    let p2sh_script_hash = hasher.finalize();
    
    let mut ripemd_hasher = Ripemd160::new();
    ripemd_hasher.update(p2sh_script_hash);
    let p2sh_hash = ripemd_hasher.finalize().to_vec();
    let mut address_bytes = vec![0x05]; // P2SH version byte
    address_bytes.extend_from_slice(&p2sh_hash);
    let mut hasher = Sha256::new();
    hasher.update(&address_bytes);
    let checksum = &Sha256::digest(&hasher.finalize())[0..4];
    address_bytes.extend_from_slice(checksum);
    let p2sh_address = bs58::encode(address_bytes).into_string();
    println!("P2SH Address: {}", p2sh_address);


    // Craft the unsigned transaction
    let mut unsigned_tx = Vec::new();

    // Transaction version (4 bytes, little-endian)
    unsigned_tx.extend_from_slice(&(1u32).to_le_bytes());

    // Input count (1 byte)
    unsigned_tx.push(1);

    // Input details
    // Previous transaction ID (all zeros)
    unsigned_tx.extend_from_slice(&[0u8; 32]);
    // Previous output index
    unsigned_tx.extend_from_slice(&(0u32).to_le_bytes());
    // ScriptSig length (initially empty)
    unsigned_tx.push(0);
    // Sequence
    unsigned_tx.extend_from_slice(&(0xFFFFFFFFu32).to_le_bytes());

    // Output count (1 byte)
    unsigned_tx.push(1);

    // Output details - 0.001 BTC to 325UUecEQuyrTd28Xs2hvAxdAjHM7XzqVF
    unsigned_tx.extend_from_slice(&(100000u64).to_le_bytes()); // 0.001 BTC in satoshis

    // Decode the address to get the script
    let address = "325UUecEQuyrTd28Xs2hvAxdAjHM7XzqVF";
    let addr_bytes = bs58::decode(address).into_vec()?;
    let script_hash = &addr_bytes[1..21]; // Extract 20-byte hash correctly

    // Create output script: OP_HASH160 <20-byte-hash> OP_EQUAL
    let mut output_script = Vec::new();
    output_script.push(0xA9); // OP_HASH160
    output_script.push(0x14); // Push 20 bytes
    output_script.extend_from_slice(script_hash);
    output_script.push(0x87); // OP_EQUAL

    // Script length
    unsigned_tx.push(output_script.len() as u8);
    // Script bytes
    unsigned_tx.extend_from_slice(&output_script);

    // Locktime (4 bytes, little-endian)
    unsigned_tx.extend_from_slice(&(0u32).to_le_bytes());

    // Calculate the sighash
    // For P2SH-P2WSH, we need to include the redeem script in the signature hash calculation
    
    // Double hash prevouts
    let mut prevouts_preimage = Vec::new();
    prevouts_preimage.extend_from_slice(&[0u8; 32]); // txid (all zeros)
    prevouts_preimage.extend_from_slice(&(0u32).to_le_bytes()); // vout
    
    let mut hasher = Sha256::new();
    hasher.update(&prevouts_preimage);
    let hash_prevouts_1 = hasher.finalize();
    
    let mut hasher = Sha256::new();
    hasher.update(&hash_prevouts_1);
    let hash_prevouts = hasher.finalize();

    // Double hash sequence
    let sequence_preimage = (0xFFFFFFFFu32).to_le_bytes();
    let mut hasher = Sha256::new();
    hasher.update(&sequence_preimage);
    let hash_sequence_1 = hasher.finalize();
    
    let mut hasher = Sha256::new();
    hasher.update(&hash_sequence_1);
    let hash_sequence = hasher.finalize();

    // Double hash outputs
    let mut outputs_preimage = Vec::new();
    outputs_preimage.extend_from_slice(&(100000u64).to_le_bytes());
    outputs_preimage.push(output_script.len() as u8);
    outputs_preimage.extend_from_slice(&output_script);
    
    let mut hasher = Sha256::new();
    hasher.update(&outputs_preimage);
    let hash_outputs_1 = hasher.finalize();
    
    let mut hasher = Sha256::new();
    hasher.update(&hash_outputs_1);
    let hash_outputs = hasher.finalize();

    // BIP143 signature hash calculation
    let mut sighash_preimage = Vec::new();
    sighash_preimage.extend_from_slice(&(1u32).to_le_bytes()); // version
    sighash_preimage.extend_from_slice(&hash_prevouts); // hash prevouts
    sighash_preimage.extend_from_slice(&hash_sequence); // hash sequence
    sighash_preimage.extend_from_slice(&[0u8; 32]); // outpoint txid (all zeros)
    sighash_preimage.extend_from_slice(&(0u32).to_le_bytes()); // outpoint index
    
    // Push the redeem script with its length prefix
    sighash_preimage.push(redeem_script.len() as u8);
    sighash_preimage.extend_from_slice(&redeem_script);
    
    // Amount being spent (0 for this exercise)
    sighash_preimage.extend_from_slice(&(0u64).to_le_bytes());
    
    // Sequence
    sighash_preimage.extend_from_slice(&(0xFFFFFFFFu32).to_le_bytes());
    
    // Hash outputs
    sighash_preimage.extend_from_slice(&hash_outputs);
    
    // Locktime
    sighash_preimage.extend_from_slice(&(0u32).to_le_bytes());
    
    // Sighash type (SIGHASH_ALL)
    sighash_preimage.extend_from_slice(&(1u32).to_le_bytes());

    // Double hash the preimage to get the sighash
    let mut hasher = Sha256::new();
    hasher.update(&sighash_preimage);
    let sighash_1 = hasher.finalize();
    
    let mut hasher = Sha256::new();
    hasher.update(&sighash_1);
    let sighash = hasher.finalize();

    // Sign the transaction with both private keys
    let secp = Secp256k1::new();
    let secret_key1 = SecretKey::from_slice(&hex::decode(priv_key1)?)?;
    let secret_key2 = SecretKey::from_slice(&hex::decode(priv_key2)?)?;
    
    let message = Message::from_slice(&sighash)?;
    let sig1 = secp.sign_ecdsa(&message, &secret_key1);
    let sig2 = secp.sign_ecdsa(&message, &secret_key2);

    // Create ECDSA signatures with SIGHASH_ALL appended
    let mut sig1_der = sig1.serialize_der().to_vec();
    sig1_der.push(0x01); // SIGHASH_ALL
    
    let mut sig2_der = sig2.serialize_der().to_vec();
    sig2_der.push(0x01); // SIGHASH_ALL

    // Construct the final transaction
    let mut final_tx = Vec::new();
    
    // Version
    final_tx.extend_from_slice(&(1u32).to_le_bytes());
    
    // Marker (0x00) and flag (0x01) for SegWit
    final_tx.push(0x00);
    final_tx.push(0x01);
    
    // Input count
    final_tx.push(1);
    
    // Input details
    final_tx.extend_from_slice(&[0u8; 32]); // txid (all zeros)
    final_tx.extend_from_slice(&(0u32).to_le_bytes()); // vout
    
    // The scriptsig for P2SH-P2WSH is just the P2SH redeem script
    let script_sig_len = p2sh_redeem_script.len() as u8;
    final_tx.push(script_sig_len);
    final_tx.extend_from_slice(&p2sh_redeem_script);
    
    // Sequence
    final_tx.extend_from_slice(&(0xFFFFFFFFu32).to_le_bytes());
    
    // Output count
    final_tx.push(1);
    
    // Output details
    final_tx.extend_from_slice(&(100000u64).to_le_bytes()); // 0.001 BTC in satoshis
    final_tx.push(output_script.len() as u8);
    final_tx.extend_from_slice(&output_script);
    
    // Witness
    // Witness item count (4 items)
    final_tx.push(4);
    
    // Empty byte for OP_CHECKMULTISIG bug
    final_tx.push(0); // Length 0
    
    // First signature (sig2)
    final_tx.push(sig2_der.len() as u8);
    final_tx.extend_from_slice(&sig2_der);
    
    // Second signature (sig1)
    final_tx.push(sig1_der.len() as u8);
    final_tx.extend_from_slice(&sig1_der);
    
    // Redeem script
    final_tx.push(redeem_script.len() as u8);
    final_tx.extend_from_slice(&redeem_script);
    
    // Locktime
    final_tx.extend_from_slice(&(0u32).to_le_bytes());

    // Convert final transaction to hex and save to out.txt
    let tx_hex = hex::encode(&final_tx);
    println!("{}", tx_hex);
    
    let mut file = File::create("out.txt")?;
    file.write_all(tx_hex.as_bytes())?;
    
    Ok(())
}